### PR TITLE
ADBDEV-5477: Invent adb_hba_file_rules_view view to show the content of pg_hba.conf

### PR DESCRIPTION
**Pull request by RekGRpth**
This PR should be merged by rebase as four separate commits.

Commit 1 refactors hba.c by by merging three separate lists into a single list
of structures. This is a full backport of 350cb92.

Commit 2 introduces a helper function strlist_to_textarray(), which is needed
for commit 3. This is a partial backport of a676201, of which the
strlist_to_textarray() function is only included.

Commit 3 adds pg_hba_file_rules() function to display the contents of
pg_hba.conf which shows the current state of the file. This is a partial
backport of de16ab7, excluding updates related to catalog changes and SSL
settings.

Commit 4 exposes pg_hba_file_rules() function to SQL through
arenadata_toolkit.adb_hba_file_rules() with a corresponding view
adb_hba_file_rules_view and adds tests.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/70436

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1369336

File: gpcontrib/arenadata_toolkit/sql/adb_hba_file_rules_view_test.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+CREATE EXTENSION arenadata_toolkit;
+SET search_path = arenadata_toolkit;
+
+-- There will surely be at least one rule
+SELECT count(*) > 0 AS ok FROM adb_hba_file_rules_view;
```
Comment (bandetto) [inline gpcontrib/arenadata_toolkit/sql/adb_hba_file_rules_view_test.sql]: Maybe for the time of this test we can backup the original pg_hba.conf, create a temporary config filled with different lines and check for expected output?

File: gpcontrib/arenadata_toolkit/arenadata_toolkit--1.5--1.6.sql [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    user_name text[],
+    address text,
+    netmask text,
+    auth_method text,
+    options text[],
+    error text
+) AS '$libdir/arenadata_toolkit', 'adb_hba_file_rules' LANGUAGE C VOLATILE STRICT;
+
+CREATE VIEW arenadata_toolkit.adb_hba_file_rules_view AS
+    SELECT * FROM arenadata_toolkit.adb_hba_file_rules() AS A;
```
Comment (bandetto) [inline gpcontrib/arenadata_toolkit/arenadata_toolkit--1.5--1.6.sql]:
```suggestion
    SELECT * FROM arenadata_toolkit.adb_hba_file_rules();
```

**Review by bandetto — COMMENTED** [review]
Can you please specify what was changed/taken from the cherry-picked commits near to "partial backport" mark in the commit messages?

File: gpcontrib/arenadata_toolkit/arenadata_toolkit--1.5--1.6.sql (hunk 2 above)
Comment (RekGRpth) [reply gpcontrib/arenadata_toolkit/arenadata_toolkit--1.5--1.6.sql]: applied

**Comment by RekGRpth** [comment]
> Can you please specify what was changed/taken from the cherry-picked commits near to "partial backport" mark in the commit messages?

added

File: gpcontrib/arenadata_toolkit/sql/adb_hba_file_rules_view_test.sql (hunk 1 above)
Comment (RekGRpth) [reply gpcontrib/arenadata_toolkit/sql/adb_hba_file_rules_view_test.sql]: done

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/70820

**Comment by bandetto** [comment]
Can you please change the PR description to:

```
This PR should be merged by rebase as four separate commits.

Commit 1 refactors hba.c by by merging three separate lists into a single list
of structures. This is a full backport of 350cb92.

Commit 2 introduces a helper function strlist_to_textarray(), which is needed
for commit 3. This is a partial backport of a676201, of which the
strlist_to_textarray() function is only included.

Commit 3 adds pg_hba_file_rules() function to display the contents of
pg_hba.conf which shows the current state of the file. This is a partial
backport of de16ab7, excluding updates related to catalog changes and SSL
settings.

Commit 4 exposes pg_hba_file_rules() function to SQL through
arenadata_toolkit.adb_hba_file_rules() with a corresponding view
adb_hba_file_rules_view and adds tests.
```

**Comment by RekGRpth** [comment]
> Can you please change the PR description to:
> 
> ```
> This PR should be merged by rebase as four separate commits.
> 
> Commit 1 refactors hba.c by by merging three separate lists into a single list
> of structures. This is a full backport of 350cb92.
> 
> Commit 2 introduces a helper function strlist_to_textarray(), which is needed
> for commit 3. This is a partial backport of a676201, of which the
> strlist_to_textarray() function is only included.
> 
> Commit 3 adds pg_hba_file_rules() function to display the contents of
> pg_hba.conf which shows the current state of the file. This is a partial
> backport of de16ab7, excluding updates related to catalog changes and SSL
> settings.
> 
> Commit 4 exposes pg_hba_file_rules() function to SQL through
> arenadata_toolkit.adb_hba_file_rules() with a corresponding view
> adb_hba_file_rules_view and adds tests.
> ```

done

**Review by bandetto — APPROVED** [review]

**Review by red1452 — COMMENTED** [review]
Why did you not add changes from [the third commit](https://github.com/arenadata/gpdb/commit/de16ab7238888b16825ad13f0bbe123632915e9b#diff-188ac8a0c4aab6cd8a532933b6d21d36d4b2d9943ed4ecdf0c7e4f3d5fa714eeR3079-R3080)? with small difference it allows to execute `pg_hba_file_rules` without adding new function to `arenadata_toolkit.c`:
```
postgres=# SELECT * FROM arenadata_toolkit.adb_hba_file_rules();
 line_number | type  |   database    | user_name |          address          |                 netmask                 | auth_method | options | error 
-------------+-------+---------------+-----------+---------------------------+-----------------------------------------+-------------+---------+-------
          86 | local | {all}         | {evgeniy} |                           |                                         | peer        |         | 
          87 | host  | {all}         | {evgeniy} | 127.0.0.1                 | 255.255.255.240                         | trust       |         | 
          88 | host  | {all}         | {evgeniy} | 192.168.0.144             | 255.255.255.255                         | trust       |         | 
          89 | host  | {all}         | {evgeniy} | 192.168.122.1             | 255.255.255.255                         | trust       |         | 
          90 | host  | {all}         | {evgeniy} | 172.17.0.1                | 255.255.255.255                         | trust       |         | 
          91 | host  | {all}         | {evgeniy} | 172.18.0.1                | 255.255.255.255                         | trust       |         | 
          92 | host  | {all}         | {evgeniy} | 192.168.124.10            | 255.255.255.255                         | trust       |         | 
          93 | host  | {all}         | {evgeniy} | 192.168.0.144             | 255.255.255.255                         | trust       |         | 
          94 | host  | {all}         | {evgeniy} | 192.168.122.1             | 255.255.255.255                         | trust       |         | 
          95 | host  | {all}         | {evgeniy} | 172.17.0.1                | 255.255.255.255                         | trust       |         | 
          96 | host  | {all}         | {evgeniy} | 172.18.0.1                | 255.255.255.255                         | trust       |         | 
          97 | host  | {all}         | {evgeniy} | 192.168.124.10            | 255.255.255.255                         | trust       |         | 
          98 | host  | {all}         | {evgeniy} | ::1                       | ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff | trust       |         | 
          99 | host  | {all}         | {evgeniy} | fe80::3ca9:9ceb:36fd:8176 | ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff | trust       |         | 
         100 | host  | {all}         | {evgeniy} | fe80::42:b6ff:fe74:b7fb   | ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff | trust       |         | 
         101 | host  | {all}         | {evgeniy} | fe80::c5e7:94db:cc6:6f7f  | ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff | trust       |         | 
         102 | host  | {all}         | {evgeniy} | fe80::18b3:66ff:fe40:c42a | ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff | trust       |         | 
         103 | local | {replication} | {evgeniy} |                           |                                         | peer        |         | 
         104 | host  | {replication} | {evgeniy} | samehost                  |                                         | trust       |         | 
         105 | host  | {replication} | {evgeniy} | 192.168.0.144             | 255.255.255.255                         | trust       |         | 
         106 | host  | {replication} | {evgeniy} | 192.168.122.1             | 255.255.255.255                         | trust       |         | 
         107 | host  | {replication} | {evgeniy} | 172.17.0.1                | 255.255.255.255                         | trust       |         | 
         108 | host  | {replication} | {evgeniy} | 172.18.0.1                | 255.255.255.255                         | trust       |         | 
         109 | host  | {replication} | {evgeniy} | 192.168.124.10            | 255.255.255.255                         | trust       |         | 
(24 rows)

postgres=# SELECT * FROM pg_hba_file_rules();
 line_number | type  |   database    | user_name |          address          |                 netmask                 | auth_method | options | error 
-------------+-------+---------------+-----------+---------------------------+-----------------------------------------+-------------+---------+-------
          86 | local | {all}         | {evgeniy} |                           |                                         | peer        |         | 
          87 | host  | {all}         | {evgeniy} | 127.0.0.1                 | 255.255.255.240                         | trust       |         | 
          88 | host  | {all}         | {evgeniy} | 192.168.0.144             | 255.255.255.255                         | trust       |         | 
          89 | host  | {all}         | {evgeniy} | 192.168.122.1             | 255.255.255.255                         | trust       |         | 
          90 | host  | {all}         | {evgeniy} | 172.17.0.1                | 255.255.255.255                         | trust       |         | 
          91 | host  | {all}         | {evgeniy} | 172.18.0.1                | 255.255.255.255                         | trust       |         | 
          92 | host  | {all}         | {evgeniy} | 192.168.124.10            | 255.255.255.255                         | trust       |         | 
          93 | host  | {all}         | {evgeniy} | 192.168.0.144             | 255.255.255.255                         | trust       |         | 
          94 | host  | {all}         | {evgeniy} | 192.168.122.1             | 255.255.255.255                         | trust       |         | 
          95 | host  | {all}         | {evgeniy} | 172.17.0.1                | 255.255.255.255                         | trust       |         | 
          96 | host  | {all}         | {evgeniy} | 172.18.0.1                | 255.255.255.255                         | trust       |         | 
          97 | host  | {all}         | {evgeniy} | 192.168.124.10            | 255.255.255.255                         | trust       |         | 
          98 | host  | {all}         | {evgeniy} | ::1                       | ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff | trust       |         | 
          99 | host  | {all}         | {evgeniy} | fe80::3ca9:9ceb:36fd:8176 | ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff | trust       |         | 
         100 | host  | {all}         | {evgeniy} | fe80::42:b6ff:fe74:b7fb   | ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff | trust       |         | 
         101 | host  | {all}         | {evgeniy} | fe80::c5e7:94db:cc6:6f7f  | ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff | trust       |         | 
         102 | host  | {all}         | {evgeniy} | fe80::18b3:66ff:fe40:c42a | ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff | trust       |         | 
         103 | local | {replication} | {evgeniy} |                           |                                         | peer        |         | 
         104 | host  | {replication} | {evgeniy} | samehost                  |                                         | trust       |         | 
         105 | host  | {replication} | {evgeniy} | 192.168.0.144             | 255.255.255.255                         | trust       |         | 
         106 | host  | {replication} | {evgeniy} | 192.168.122.1             | 255.255.255.255                         | trust       |         | 
         107 | host  | {replication} | {evgeniy} | 172.17.0.1                | 255.255.255.255                         | trust       |         | 
         108 | host  | {replication} | {evgeniy} | 172.18.0.1                | 255.255.255.255                         | trust       |         | 
         109 | host  | {replication} | {evgeniy} | 192.168.124.10            | 255.255.255.255                         | trust       |         | 
(24 row)
```
You should add this lines to the `src/include/catalog/pg_proc.h` file:
```diff
diff --git a/src/include/catalog/pg_proc.h b/src/include/catalog/pg_proc.h
index 98f85d01ba0..afabc920ad5 100644
--- a/src/include/catalog/pg_proc.h
+++ b/src/include/catalog/pg_proc.h
@@ -3113,6 +3113,8 @@ DESCR("SET X as a function");
 DATA(insert OID = 2084 (  pg_show_all_settings PGNSP PGUID 12 1 1000 0 0 f f f f t t s 0 0 2249 "" "{25,25,25,25,25,25,25,25,25,25,25,1009,25,25,25,23}" "{o,o,o,o,o,o,o,o,o,o,o,o,o,o,o,o}" "{name,setting,unit,category,short_desc,extra_desc,context,vartype,source,min_val,max_val,enumvals,boot_val,reset_val,sourcefile,sourceline}" _null_ show_all_settings _null_ _null_ _null_ ));
 DESCR("SHOW ALL as a function");
 GPDB_EXTRA_COL(prodataaccess = r);
+DATA(insert OID = 3401 (  pg_hba_file_rules		PGNSP PGUID 12 1 1000 0 0 f f f f t t s 0 0 2249 "" "{23,25,1009,1009,25,25,25,1009,25}" "{o,o,o,o,o,o,o,o,o}" "{line_number,type,database,user_name,address,netmask,auth_method,options,error}" _null_ pg_hba_file_rules _null_ _null_ _null_ ));
+DESCR("show pg_hba.conf rules");
 /* A bit different on GPDB */
 DATA(insert OID = 1371 (  pg_lock_status   PGNSP PGUID 12 1 1000 0 0 f f f f t t v 0 0 2249 "" "{25,26,26,23,21,25,28,26,26,21,25,23,25,16,16,23,16,23}" "{o,o,o,o,o,o,o,o,o,o,o,o,o,o,o,o,o,o}" "{locktype,database,relation,page,tuple,virtualxid,transactionid,classid,objid,objsubid,virtualtransaction,pid,mode,granted,fastpath,mppsessionid,mppiswriter,gp_segment_id}" _null_ pg_lock_status _null_ _null_ _null_ ));
 DESCR("view system lock information");
```

**Comment by RekGRpth** [comment]
> You should add this lines to the `src/include/catalog/pg_proc.h` file:

This is a catalog update that is not allowed! Who will update the catalog for customers?

File: gpcontrib/arenadata_toolkit/expected/adb_hba_file_rules_view_test.out [hunk 3]
```diff
@@ -0,0 +1,17 @@
+CREATE EXTENSION arenadata_toolkit;
+SET search_path = arenadata_toolkit;
+-- save pg_hba.conf
+\! mv -f $MASTER_DATA_DIRECTORY/pg_hba.conf $MASTER_DATA_DIRECTORY/pg_hba.conf.back
+-- fill pg_hba.conf with test entry
+\! echo "local    all    all    trust" >> $MASTER_DATA_DIRECTORY/pg_hba.conf
```
Comment (red1452) [inline gpcontrib/arenadata_toolkit/expected/adb_hba_file_rules_view_test.out:6]: If you do not want to see something other at the `SELECT`, change `>>` to `>`.

**Review by red1452 — APPROVED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/71235

**Pull request merged**